internal/unpack/bi_reader.go

```diff
@@ -0,0 +1,26 @@
+package unpack
+
+import "io"
+
+// biReader is a specialized io.MultiReader optimized for
+// only two readers
+type biReader struct {
+	readers [2]io.Reader
+	cur     uint8
+}
+
+func (mr *biReader) Read(p []byte) (n int, err error) {
+	// this local is important for bounds-check-elimination
+	cur := mr.cur
+	if cur < 2 {
+		n, err = mr.readers[cur].Read(p)
+		if err == io.EOF {
+			mr.readers[cur] = nil
+			mr.cur++
+		}
+		if mr.cur == 1 && err == io.EOF {
+			err = nil
+		}
+	}
```

Comment:
```suggestion
	if cur >= 2 {
		return 0, io.EOF
	}
	n, err = mr.readers[cur].Read(p)
	if err == io.EOF {
		mr.readers[cur] = nil
		mr.cur++
		if mr.cur == 1 {
			err = nil
		}
	}
	return
}
```

Don't you need to ensure that `io.EOF` is returned on all subsequent calls?

Reply:
Probably yes lol, this was specifically the area I wanted eyes on coz its so damn fiddly